assertion failed: Inconsistent state in TS[212X, 211, 2, 1, 0]: it owns TypeVar(TypeParamRef(S2)) whose owningState is TS[211, 2, 1, 0]

Maybe related to #13101 but I can repro in latest RC, which I think has a fix for that.

## Compiler version

3.0.1, 3.0.2-RC1, 3.0.2-RC2

## Minimized code

```Scala
sealed trait Assertion[Cond <: Boolean]

object Assertion {
  given Assertion[true] with {
  }
}


trait Tensor[S <: Tuple] {
  // === isn't defined
  def +[S2 <: Tuple](other: Tensor[S2])(using Assertion[S === S2]): Tensor[S] = new Tensor {}
}

object Tensor {
  def mk[S <: Tuple]: Tensor[S] = new Tensor {}
}

object Foo {
  val t1: Tensor[("batch", "len", "embed")] = Tensor.mk
  val t2: Tensor[("embed", "hid")] = Tensor.mk
  def foo(x: Any) = x
  // ok
//  foo(t1 + t2)

  def bar(x: Any) = {
    x
  }

  bar ( foo (t1 +t2))
}
```

## Output (click arrow to expand)
<details>

```scala
[error] java.lang.AssertionError: assertion failed: Inconsistent state in TS[212X, 211, 2, 1, 0]: it owns TypeVar(TypeParamRef(S2)) whose owningState is TS[211, 2, 1, 0]
[error] 	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] 	at dotty.tools.dotc.core.TyperState.gc$$anonfun$1(TyperState.scala:220)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] 	at dotty.tools.dotc.util.SimpleIdentitySet$Set1.foreach(SimpleIdentitySet.scala:69)
[error] 	at dotty.tools.dotc.core.TyperState.gc(TyperState.scala:226)
[error] 	at dotty.tools.dotc.core.TyperState.commit(TyperState.scala:163)
[error] 	at dotty.tools.dotc.typer.Inferencing$.isFullyDefined(Inferencing.scala:40)
[error] 	at dotty.tools.dotc.typer.Implicits.ignoredInstanceNormalImport$1(Implicits.scala:881)
[error] 	at dotty.tools.dotc.typer.Implicits.$anonfun$1(Implicits.scala:896)
[error] 	at dotty.tools.dotc.typer.ImplicitSearchError.hiddenImplicitsAddendum(ErrorReporting.scala:419)
[error] 	at dotty.tools.dotc.typer.ImplicitSearchError.missingArgMsg(ErrorReporting.scala:267)
[error] 	at dotty.tools.dotc.typer.Implicits.missingArgMsg(Implicits.scala:897)
[error] 	at dotty.tools.dotc.typer.Implicits.missingArgMsg$(Implicits.scala:774)
[error] 	at dotty.tools.dotc.typer.Typer.missingArgMsg(Typer.scala:106)
[error] 	at dotty.tools.dotc.typer.Typer.issueErrors$1$$anonfun$1$$anonfun$1(Typer.scala:3353)
[error] 	at dotty.tools.dotc.reporting.NoExplanation.msg(Message.scala:132)
[error] 	at dotty.tools.dotc.reporting.Message.message(Message.scala:87)
[error] 	at dotty.tools.dotc.reporting.Message.isNonSensical(Message.scala:99)
[error] 	at dotty.tools.dotc.reporting.HideNonSensicalMessages.isHidden(HideNonSensicalMessages.scala:16)
[error] 	at dotty.tools.dotc.reporting.HideNonSensicalMessages.isHidden$(HideNonSensicalMessages.scala:10)
[error] 	at dotty.tools.dotc.reporting.AbstractReporter.isHidden(AbstractReporter.scala:8)
[error] 	at dotty.tools.dotc.reporting.Reporter.report(Reporter.scala:148)
[error] 	at dotty.tools.dotc.reporting.Reporter.flush$$anonfun$1(Reporter.scala:206)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] 	at scala.collection.immutable.List.foreach(List.scala:333)
[error] 	at dotty.tools.dotc.reporting.Reporter.flush(Reporter.scala:206)
[error] 	at dotty.tools.dotc.core.TyperState.commit(TyperState.scala:143)
[error] 	at dotty.tools.dotc.typer.Applications.fail$1(Applications.scala:986)
[error] 	at dotty.tools.dotc.typer.Applications.realApply$7$$anonfun$6$$anonfun$2(Applications.scala:992)
[error] 	at scala.Option.getOrElse(Option.scala:201)
[error] 	at dotty.tools.dotc.typer.Applications.realApply$8$$anonfun$7(Applications.scala:993)
[error] 	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2952)
[error] 	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:994)
[error] 	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1032)
[error] 	at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:317)
[error] 	at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:106)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2693)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error] 	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2874)
[error] 	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2897)
[error] 	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2352)
[error] 	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2681)
[error] 	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2685)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2755)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error] 	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2847)
[error] 	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2897)
[error] 	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2475)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2726)
[error] 	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2756)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2821)
[error] 	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2825)
[error] 	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2941)
[error] 	at dotty.tools.dotc.typer.FrontEnd.liftedTree1$1(FrontEnd.scala:79)
[error] 	at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:84)
[error] 	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:43)
[error] 	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:85)
[error] 	at dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:120)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] 	at scala.collection.immutable.List.foreach(List.scala:333)
[error] 	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:120)
[error] 	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] 	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] 	at dotty.tools.dotc.Run.runPhases$5(Run.scala:216)
[error] 	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:224)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] 	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] 	at dotty.tools.dotc.Run.compileUnits(Run.scala:231)
[error] 	at dotty.tools.dotc.Run.compileSources(Run.scala:166)
[error] 	at dotty.tools.dotc.Run.compile(Run.scala:150)
[error] 	at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] 	at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] 	at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] 	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:192)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:182)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] 	at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error] 	at sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:175)
[error] 	at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:173)
[error] 	at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error] 	at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] 	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] 	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] 	at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] 	at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error] 	at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:501)
[error] 	at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error] 	at sbt.internal.inc.Incremental$.apply(Incremental.scala:167)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] 	at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2357)
[error] 	at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2314)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] 	at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] 	at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2310)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:291)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
[error] 	at java.base/java.lang.Thread.run(Thread.java:831)

```
</details>

--- a/compiler/src/dotty/tools/dotc/core/TyperState.scala
+++ b/compiler/src/dotty/tools/dotc/core/TyperState.scala
@@ -162,6 +162,7 @@ class TyperState() {
         targetState.mergeConstraintWith(this)
     targetState.gc()
     isCommitted = true
+    ownedVars = SimpleIdentitySet.empty
   }
 
   /** Ensure that this constraint does not associate different TypeVars for the

--- a/tests/neg/i13407.scala
+++ b/tests/neg/i13407.scala
@@ -1,0 +1,19 @@
+import scala.quoted.Type
+
+trait Tensor[S <: Tuple] {
+  def sum[Axis <: Shape: Type](axis: Axis): Tensor[S] = { // error
+    Tensor.mk
+  }
+}
+
+object Tensor {
+  def mk[S <: Tuple]: Tensor[S] = new Tensor {}
+}
+
+object Foo {
+  val t1: Tensor[("batch", "len", "embed")] = Tensor.mk
+   def foo(x: Any) = {
+
+  }
+  foo(foo(t1.sum("len"))) // error
+}

